### PR TITLE
Collect services

### PR DESCRIPTION
**Pull request by xlucas**
### What does this PR do?

Collect Kubernetes Services.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -201,3 +201,5 @@ replace (
 
 // Pinned so it includes fix for Windows Nano
 replace k8s.io/klog => k8s.io/klog v1.0.1-0.20200310124935-4ad0115ba9e4
+
+replace github.com/DataDog/agent-payload => github.com/DataDog/agent-payload v0.0.0-20200722131956-8c38b0f51477
```
Comment (xlucas) [inline go.mod:205]: Note: will bump this before merging to point to the latest agent-payload released including the corresponding definitions.

File: pkg/util/kubernetes/apiserver/controllers.go [hunk 2]
```diff
@@ -149,25 +149,31 @@ func startAutoscalersController(ctx ControllerContext, c chan error) {
...
+
 	// Just start the shared informer, the autodiscovery
 	// components will access it when needed.
-	go ctx.InformerFactory.Core().V1().Endpoints().Informer().Run(ctx.StopCh)
+	ctx.InformerFactory.Start(ctx.StopCh)
 
 	// Wait for the cache to sync
 	c <- SyncInformers(map[InformerName]cache.SharedInformer{
-		endpointsInformer: ctx.InformerFactory.Core().V1().Endpoints().Informer(),
+		endpointsInformer: informer,
```
Comment (xlucas) [inline pkg/util/kubernetes/apiserver/controllers.go:177]: Not quite sure about those, my understanding is that calling `Run()` directly as it was done before would not keep  track of this in the started informers map inside the informer factory and it's necessary to call `.Informer()` before calling `.Start()` so it runs `InformerFor()` and register the informer in the factory.

File: pkg/clusteragent/orchestrator/transformer.go [hunk 3]
```diff
@@ -72,6 +72,80 @@ func extractReplicaSet(rs *v1.ReplicaSet) *model.ReplicaSet {
...
+			Port:       port.Port,
+			TargetPort: port.TargetPort.String(),
+			NodePort:   port.NodePort,
+		})
+	}
+
+	return message
+}
+
+func extractServiceSelector(ls map[string]string) []*model.LabelSelectorRequirement {
```
Comment (mfpierre) [inline pkg/clusteragent/orchestrator/transformer.go:137]: They add to use a different struct than deploy/rs 😑

Comment (xlucas) [reply pkg/clusteragent/orchestrator/transformer.go:137]: ¯\_(ツ)_/¯

File: pkg/clusteragent/orchestrator/processor.go [hunk 4]
```diff
@@ -108,15 +104,10 @@ func processReplicaSetList(rsList []*v1.ReplicaSet, groupID int32, cfg *config.A
 
 		// k8s objects only have json "omitempty" annotations
 		// we're doing json<>yaml to get rid of the null properties
-		jsonRs, err := jsoniter.Marshal(rsList[rs])
-		if err != nil {
-			log.Debugf("Could not marshal replica set in JSON: %s", err)
+		if err := extractYaml(&rsModel.Yaml, rsList[rs]); err != nil {
+			log.Debugf("Could not marshal replica set into JSON: %s", err)
```
[the remark this answers is not in the document]
Comment (xlucas) [reply pkg/clusteragent/orchestrator/processor.go:108]: Yeah for now it's consistent with all other parts that live in the agent or in the cluster-agent so I'd better change it everywhere later when we release those together.

File: pkg/util/kubernetes/apiserver/controllers.go (hunk 2 above)
Comment (xlucas) [reply pkg/util/kubernetes/apiserver/controllers.go:177]: cc @DataDog/container-integrations

File: go.mod (hunk 1 above)
Comment (xlucas) [reply go.mod]: Done

File: pkg/util/kubernetes/apiserver/controllers.go [hunk 5]
```diff
@@ -149,25 +149,31 @@ func startAutoscalersController(ctx ControllerContext, c chan error) {
...
 // startEndpointsInformer starts the endpoints informer.
 // The synchronization of the endpoints informer is handled in this function.
 func startEndpointsInformer(ctx ControllerContext, c chan error) {
+
+	informer := ctx.InformerFactory.Core().V1().Endpoints().Informer()
+
 	// Just start the shared informer, the autodiscovery
 	// components will access it when needed.
-	go ctx.InformerFactory.Core().V1().Endpoints().Informer().Run(ctx.StopCh)
+	ctx.InformerFactory.Start(ctx.StopCh)
```
Comment (clamoriniere) [inline pkg/util/kubernetes/apiserver/controllers.go:173]: Is it always the same ctx.InformerFactory that we start in the different methods: `startEndpointsInformer ` `startServicesInformer`...? 
If yes, it should be done outside those functions.

Comment (xlucas) [reply pkg/util/kubernetes/apiserver/controllers.go:173]: Yes it's always the same factory. However it's also started here https://github.com/DataDog/datadog-agent/blob/cf0550629056d6a9cc90249892f92f9fff2a35af/pkg/util/kubernetes/apiserver/controllers.go#L105.

Comment (xlucas) [reply pkg/util/kubernetes/apiserver/controllers.go:173]: So I'm a bit unclear about the pattern here on those two handlers, should I just create a map of informers in the context, populate it from `startEndpointsInformer` & `startServicesInformer` then call `WaitForCacheSync` right after the common factory start above?

Comment (xlucas) [reply pkg/util/kubernetes/apiserver/controllers.go:173]: Like this: https://github.com/DataDog/datadog-agent/commit/43de23ac73523b2f92c7e257ce8039f4522b708f

File: pkg/util/kubernetes/apiserver/controllers.go (hunk 2 above)
Comment (clamoriniere) [reply pkg/util/kubernetes/apiserver/controllers.go:177]: Thanks for the clarification. Looks good to me

File: pkg/util/kubernetes/apiserver/controllers.go (hunk 5 above)
Comment (clamoriniere) [reply pkg/util/kubernetes/apiserver/controllers.go:173]: Yes

**Review by clamoriniere — APPROVED** [review]

**Review by mfpierre — APPROVED** [review]
:shipit:

File: pkg/util/kubernetes/apiserver/controllers.go [hunk 6]
```diff
@@ -46,15 +46,16 @@ var controllerCatalog = map[controllerName]controllerFuncs{
 	},
 	servicesController: {
 		func() bool { return config.Datadog.GetBool("cluster_checks.enabled") },
-		startServicesInformer,
+		registerServicesInformer,
```
Comment (nammn) [inline pkg/util/kubernetes/apiserver/controllers.go:49]: just realized. Did we want to register a own controller for this?

**Pull request merged**